src/devices/MatrixKeyboard/MatrixKeyboard.cs

```diff
@@ -0,0 +1,231 @@
+        private readonly int[] _columnPins;
+        private int _currentRow;
+        private readonly PinValue[] _buttonValues;
+        private double _scanFrequency;
+
+        /// <summary>
+        /// Initialize keyboard
+        /// </summary>
+        /// <param name="gpioController">GPIO controller</param>
+        /// <param name="rowPins">Row pins</param>
+        /// <param name="colPins">Column pins</param>
+        /// <param name="scanFreq">Scanning frequency</param>
+        public MatrixKeyboard(IGpioController gpioController, IEnumerable<int> rowPins, IEnumerable<int> colPins, double scanFreq = 50)
+        {
+            _gpio = gpioController ?? throw new ArgumentNullException(nameof(gpioController));
+
+            if (rowPins == null)
+            {
+                throw new ArgumentNullException(nameof(rowPins));
+            }
+            if (rowPins.Count() < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(rowPins), "The number of rows is at least 1");
+            }
+
+            if (colPins == null)
+            {
+                throw new ArgumentNullException(nameof(colPins));
+            }
+            if (colPins.Count() < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(colPins), "The number of columns is at least 1");
+            }
+
+            if (scanFreq <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(scanFreq), "Scanning frequency must be positive");
+            }
+
+            _rowPins = rowPins.ToArray();
+            _columnPins = colPins.ToArray();
+            _buttonValues = new PinValue[_rowPins.Length * _columnPins.Length];
+
+            _scanThread = new Thread(ScanThreadStart);
+            IsScanning = false;
+            ScanFrequency = scanFreq;
+        }
+
+        /// <summary>
+        /// Open GPIO pins then start keyboard scanning
+        /// </summary>
+        public void StartScan()
+        {
+            if (!IsScanning)
+            {
+                try
+                {
+                    for (var i = 0; i < _rowPins.Length; i++)
+                    {
+                        _gpio.OpenPin(_rowPins[i], PinMode.Output);
+                    }
+                    for (var i = 0; i < _columnPins.Length; i++)
+                    {
+                        _gpio.OpenPin(_columnPins[i], PinMode.Input);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    IsScanning = false;
+                    throw new Exception("Error while openning GPIO pins.", ex);
+                }
+
+                IsScanning = true;
+                try
+                {
+                    _scanThread.Start();
+                }
+                catch (Exception ex)
+                {
+                    IsScanning = false;
+                    throw new Exception("Error while starting scanning thread.", ex);
```

Comment:
under what circumstances can this fail? Should you just not do anything, it does not seem very common and wrapping in Exception might not be useful